Add flowchart implementation

TrueFinch opened the pull request:
Add Flowchart implementation from perl to python. 'to_svg' and 'to_svg_main' move to Prog mogule
https://www.notion.so/Add-Flowchart-module-461ce317ebb046a6b2764f21f8ab013a

TrueFinch committed:
Add flowchart implementation

https://www.notion.so/Add-Flowchart-module-461ce317ebb046a6b2764f21f8ab013a

SergeyZinkovich commented on EGE/Prog.py:
```diff
@@ -410,6 +413,45 @@ def complexity(self, env, mistakes, iter_):
+            if st and isinstance(st, Assign):
+                linear.append(st.to_lang_named('Alg'))
+            else:
+                if linear:
```
Это условие всегда истинно. В перле это было типо if len(linear):

TrueFinch replied:
> Это условие всегда истинно. В перле это было типо if len(linear):

То есть его просто убрать?

TrueFinch replied:
А вот тут ты не прав: оно ложно, когда linear пустой, а когда он пустой, мы не хотим пихать его в elements

SergeyZinkovich replied:
Действительно, в питоне это тоже проверяет размер...

SergeyZinkovich commented on EGE/ProgModules/Flowchart.py:
```diff
@@ -0,0 +1,156 @@
+        if jump_data is None:
+            jump_data = Jump()
+        self.jumps_.append(jump_data)
+        return self.jumps_[len(self.jumps_) - 1]
```
А почему бы не возвращать jump_data? Ну лили на крайняк self.jumps_[-1]

TrueFinch replied:
> А почему бы не возвращать jump_data? Ну лили на крайняк self.jumps_[-1]

Ну если это точно один и тот же объект будет, то можно. Не был до конца в этом уверен

TrueFinch replied:
Сейчас почекаю

TrueFinch replied:
Истинно так

TrueFinch committed:
Simplify return value

https://www.notion.so/Add-Flowchart-module-461ce317ebb046a6b2764f21f8ab013a — according to https://github.com/klenin/EGEpy/pull/25#discussion_r784856182

SergeyZinkovich commented on EGE/GenBase.py:
```diff
@@ -1,12 +1,14 @@
+    rnd: EGE.Random.Random
 
-    def __init__(self, rnd, text: str = None, correct = None):
+    def __init__(self, rnd: EGE.Random.Random, text: str = None, correct = None):
```
Надо решить, ставим ли мы тайпхинты на классы

TrueFinch replied:
Я за то, чтобы их ставить: так процесс написания кода становится намного приятнее, потому что среда понимает, у кого какой метод ты пытаешься вызвать и может что-то подсказать
(просто перенёс то, что писал по этому поводу в дискорде)

tonichoff replied:
Согласен с Вовой

tonichoff commented on EGE/Prog.py:
```diff
@@ -410,6 +413,45 @@ def complexity(self, env, mistakes, iter_):
+        r = ''
+        elements = []
+        linear = []
+        for st in self.statements + [None]:
```
Нужны пробелы внутри списка вокруг ключевого слова `None`

tonichoff commented on EGE/Prog.py:
```diff
@@ -410,6 +413,45 @@ def complexity(self, env, mistakes, iter_):
+        r = svg.g(f"\n{r}" + f.jumps(),
+                  stroke='black',
+                  fill='none'
+                  ) + f.texts()
```
А такой перенос аргументов описан в PEP8? Он мне кажется странным

TrueFinch replied:
Ну согласно вот этой главе https://www.python.org/dev/peps/pep-0008/#code-lay-out кажется, что действительно не очень верно. Но там всё равно нет однозначного ответа, как надо

tonichoff commented on EGE/Prog.py:
```diff
@@ -590,6 +632,28 @@ def run(self, env):
+        r = f.add_rhomb(
+            self.cond.to_lang_named('Alg'),
+            enter, { 'right': exit_, 'middle': middle }
+        )
```
А тут как раз у меня нет вопросов к переносу аргументов вызова функции

tonichoff commented on EGE/ProgModules/Flowchart.py:
```diff
@@ -0,0 +1,156 @@
+        self.y2 += step
+
+    def point(self, x: float = None, y: float = None):
+        return {'x': 0 if x is None else x,
```
Пробелов не хватает после открывающей и до закрывающей фигурных скобок

tonichoff commented on EGE/ProgModules/Flowchart.py:
```diff
@@ -0,0 +1,156 @@
+            else:
+                p += 'v10 '
+            p += f"H{x} V{dy} H{dx} {ah}"
+        return svg.path(d=p)
```
Не хватает пустой строки в конце файла

ToriRori replied:
гитхаб спрятал

tonichoff commented on EGE/ProgModules/Flowchart.py:
```diff
@@ -0,0 +1,156 @@
+                    { 'x': sx, 'y': sy - FONT_SZ / 2 }
+                )
+            ah = arrow_head('down' if dy > sy else
+                                 'left' if right else 'right')
```
Смущает форматирование вложенного тернарного оператора. Принятие решения вносить изменение или нет отдаю ревьюемому

TrueFinch replied:
Меня вообще бесит то, как тернарники выглядят в питоне

TrueFinch committed:
Fix code according to comment from pull requests

https://www.notion.so/Add-Flowchart-module-461ce317ebb046a6b2764f21f8ab013a — https://github.com/klenin/EGEpy/pull/25#discussion_r785402199, https://github.com/klenin/EGEpy/pull/25#discussion_r785402293, https://github.com/klenin/EGEpy/pull/25#discussion_r785402351, https://github.com/klenin/EGEpy/pull/25#discussion_r785402512, https://github.com/klenin/EGEpy/pull/25#discussion_r785402559, https://github.com/klenin/EGEpy/pull/25#discussion_r785402754
